include/gaussian_samplers.h

```diff
@@ -0,0 +1,365 @@
+
+
+#ifndef GAUSSIAN_SAMPLERS_H
+#define GAUSSIAN_SAMPLERS_H
+
+//#include <stdio.h>
+//#include <math.h>
+
+NT eval_exp(Point p, NT a){
+    return std::exp(-a*p.squared_length());
+}
+
+
+NT get_max(Point l, Point u, NT a_i){
+    NT res;
+    Point a = -1.0*l;
+    Point bef = u-l;
+    Point b = (1.0/std::sqrt((bef).squared_length()))*bef;
+    Point z = (a.dot(b)*b)+l;
+    NT low_bd = (l[0]-z[0])/b[0];
+    NT up_bd = (u[0]-z[0])/b[0];
+    if(low_bd*up_bd>0){
+    //if(std::signbit(low_bd)==std::signbit(up_bd)){
+        res = std::max(eval_exp(u,a_i),eval_exp(l,a_i));
+    }else{
+        res = eval_exp(z,a_i);
+    }
+
+    return res;
+}
+
+
+NT get_max_coord(NT l, NT u, NT a_i){
+    if(l<0.0 && u>0.0){
+        return 1.0;
+    }
+    return std::max(std::exp(-a_i*l*l), std::exp(-a_i*u*u));
+}
+
+
+int rand_exp_range(Point lower, Point upper, NT a_i, Point &p, vars_g &var){
+    NT r, r_val, fn;
+    Point bef = upper-lower;
+    //unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+    //RNGType rng(seed);
+    RNGType &rng2 = var.rng;
+    if(a_i>0.00000001 && std::sqrt(bef.squared_length()) >= (2.0/std::sqrt(2.0*a_i))){
+        boost::normal_distribution<> rdist(0,1);
+        Point a = -1.0*lower;
+        Point b = (1.0/std::sqrt(bef.squared_length()))*bef;
+        Point z = (a.dot(b)*b)+lower;
+        NT low_bd = (lower[0]-z[0])/b[0];
+        NT up_bd = (upper[0]-z[0])/b[0];
+        while(true){
+            r = rdist(rng2);
+            r = r/std::sqrt(2.0*a_i);
+            if(r>=low_bd && r<=up_bd){
+                break;
+            }
+        }
+        p=(r*b)+z;
+
+    }else{
+        boost::random::uniform_real_distribution<> urdist(0,1);
+        NT M=get_max(lower, upper, a_i);
+        while(true){
+            r=urdist(rng2);
+            Point pef = r*upper;
+            p = ((1.0-r)*lower)+ pef;
+            r_val = M*urdist(var.rng);
+            fn = eval_exp(p,a_i);
+            if(r_val<fn){
+                break;
+            }
+        }
+    }
+    return 1;
+}
+
+
+int rand_exp_range_coord(NT l, NT u, NT a_i, NT &dis, vars_g &var){
+    NT r, r_val, fn;
+    //unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+    //RNGType rng(seed);
+    RNGType &rng2 = var.rng;
+    if(a_i>std::pow(10,-8.0) && u-l>=2.0/std::sqrt(2.0*a_i)){
+        boost::normal_distribution<> rdist(0,1);
+        while(true){
+            r = rdist(rng2);
+            r = r/std::sqrt(2.0*a_i);
+            if(r>=l && r<=u){
+                break;
+            }
+        }
+        dis=r;
+
+    }else{
+        boost::random::uniform_real_distribution<> urdist(0,1);
+        NT M = get_max_coord(l, u, a_i);
+        while(true){
+            r=urdist(rng2);
+            dis = (1.0-r)*l+r*u;
+            r_val = M*urdist(rng2);
+            fn = std::exp(-a_i*dis*dis);
+            if(r_val<fn){
+                break;
+            }
+        }
+    }
+
+    return 1;
+}
+
```

Comment:
there are many similarities between ` rand_exp_range_coord` and `rand_exp_range` consider to use function to avoid duplicate code